File: apps/ios/ScribbleFit/Sources/Features/Canvas/CanvasViewModel.swift
import Combine
import Foundation

public struct CanvasUiState: Sendable {
    public var greeting: String = ""
    public var userName: String = "George"
    public var scribbleText: String = ""
    public var scribbles: [Scribble] = []
    public var isSyncing: Bool = false
    public var isRecording: Bool = false
}

@MainActor
public final class CanvasViewModel: ObservableObject {
    @Published public var uiState = CanvasUiState()

    private let scribbleRepository: any ScribbleRepository
    private let confirmWorkoutUseCase: ConfirmWorkoutUseCase
    private var cancellables = Set<AnyCancellable>()

    public init(
        scribbleRepository: any ScribbleRepository,
        confirmWorkoutUseCase: ConfirmWorkoutUseCase
    ) {
        self.scribbleRepository = scribbleRepository
        self.confirmWorkoutUseCase = confirmWorkoutUseCase

        scribbleRepository.getAllScribbles()
            .receive(on: DispatchQueue.main)
            .sink { [weak self] scribbles in
                self?.uiState.scribbles = scribbles
                self?.uiState.isSyncing = scribbles.contains {
                    if case .raw(_, _, _, let status) = $0 { return status == .pending }
                    return false
                }
            }
            .store(in: &cancellables)

        uiState.greeting = greeting()
    }

    public func onTextChange(_ newText: String) {
        uiState.scribbleText = newText
    }

    public func submitScribble() {
        let text = uiState.scribbleText.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !text.isEmpty else { return }
        Task {
            await scribbleRepository.enqueueScribble(rawText: text)
            uiState.scribbleText = ""
        }
    }

    public func onRetryScribble(id: String) {
        Task { try? await scribbleRepository.updateSyncStatus(id: id, status: .pending) }
    }

    public func onConfirmClick(parsedWorkout: ParsedWorkout, scribbleId: String) {
        Task { try? await confirmWorkoutUseCase.execute(parsedWorkout: parsedWorkout, scribbleId: scribbleId) }
    }

    public func onMicClick() {
        uiState.isRecording.toggle()
    }

    private func greeting() -> String {
        let hour = Calendar.current.component(.hour, from: Date())
        switch hour {
        case 0..<12: return "Good morning"
        case 12..<17: return "Good afternoon"
        default: return "Good evening"
        }
    }
}

File: apps/android/feature/canvas/ui/src/main/java/com/scribblefit/feature/canvas/ui/CanvasViewModel.kt
package com.scribblefit.feature.canvas.ui

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.scribblefit.feature.ai.domain.model.AnalysisSuggestion
import com.scribblefit.feature.canvas.domain.usecase.ConfirmWorkoutUseCase
import com.scribblefit.feature.canvas.domain.usecase.ProcessScribbleUseCase
import com.scribblefit.feature.scribble.domain.Scribble
import com.scribblefit.feature.scribble.domain.ScribbleRepository
import com.scribblefit.feature.scribble.domain.SyncScribblesUseCase
import com.scribblefit.feature.scribble.domain.SyncStatus
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.SharingStarted
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.combine
import kotlinx.coroutines.flow.stateIn
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch
import java.util.Calendar
import javax.inject.Inject

private const val FLOW_TIMEOUT_MS = 5_000L

data class CanvasUiState(
    val greeting: String = "",
    val userName: String = "George",
    val homeSuggestion: AnalysisSuggestion? = null,
    val scribbleText: String = "",
    val feedItems: List<Scribble> = emptyList(),
    val isSyncing: Boolean = false,
    val isRecording: Boolean = false
)

@HiltViewModel
class CanvasViewModel @Inject constructor(
    private val scribbleRepository: ScribbleRepository,
    private val processScribbleUseCase: ProcessScribbleUseCase,
    private val confirmWorkoutUseCase: ConfirmWorkoutUseCase,
    private val syncScribbles: SyncScribblesUseCase
) : ViewModel() {

    private val _scribbleText = MutableStateFlow("")
    private val _extras = MutableStateFlow(CanvasExtras())

    val uiState: StateFlow<CanvasUiState> = combine(
        scribbleRepository.getAllScribbles(),
        _scribbleText,
        _extras
    ) { scribbles, scribbleText, extras ->
        CanvasUiState(
            greeting = getGreeting(),
            scribbleText = scribbleText,
            feedItems = scribbles,
            homeSuggestion = extras.homeSuggestion,
            isRecording = extras.isRecording,
        )
    }.stateIn(viewModelScope, SharingStarted.WhileSubscribed(FLOW_TIMEOUT_MS), CanvasUiState())

    fun onTextChange(newText: String) {
        _scribbleText.update { newText }
    }

    fun submitScribble() {
        val text = _scribbleText.value.trim()
        if (text.isBlank()) return
        viewModelScope.launch {
            processScribbleUseCase.execute(text)
            _scribbleText.update { "" }
        }
    }

    fun onRetryScribble(id: String) {
        viewModelScope.launch { scribbleRepository.updateSyncStatus(id, SyncStatus.Pending) }
    }

    fun onConfirmClick(parsed: Scribble.Parsed) {
        viewModelScope.launch { confirmWorkoutUseCase.execute(parsed) }
    }

    fun onMicClick() {
        _extras.update { it.copy(isRecording = !it.isRecording) }
    }

    private fun getGreeting(): String {
        val hour = Calendar.getInstance().get(Calendar.HOUR_OF_DAY)
        return when {
            hour < MORNING_CUTOFF -> "Good morning"
            hour < AFTERNOON_CUTOFF -> "Good afternoon"
            else -> "Good evening"
        }
    }

    private data class CanvasExtras(
        val homeSuggestion: AnalysisSuggestion? = null,
        val isRecording: Boolean = false
    )

    init {
        viewModelScope.launch {
            syncScribbles()
        }
    }

    companion object {
        private const val MORNING_CUTOFF = 12
        private const val AFTERNOON_CUTOFF = 17
    }
}
